---
layout: tutorial_hands_on
topic_name: introduction
tutorial_name: intro_NGS_analysis
---

# Introduction
{:.no_toc}

In this tutorial you will learn the basics on how to work and analyse next-generation sequencing data. First we will deal with manipulation of fastq files, more precisely how to access the quality of the data and perform some quality treatments, and then we will deal with SAM/BAM files, learning how to map the data and visualize the results.  

> ### Agenda
>
> In this tutorial, we will deal with:
>
> 1. TOC
> {:toc}
>
{: .agenda}

# Fast manipulation 

Most high-throughput sequencing machines output fastq files, the “de facto” current standard in HTS.
Fastq files are simply text files, where each block of information (a sequenced DNA fragment) in this
format is encoded as 4 lines. If you look at the first lines of a fastq file, you would see something like:

@HWI-M01876:76:000000000-AF16W:1:1101:10853:1000 1:N:0:CGTGACAGAT
NTGTACTTCATCCGAAACTCGTGCTCATCTCTGCTCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCGTGAT
+
#8ABCFGGGFCEDCFGGGGGGGFFCGEFGGGGGGFGGGGGGGGDEFGGGGGGGGGGGGGGGGGFFFEGGGGGGGGF
@HWI-M01876:76:000000000-AF16W:1:1101:16471:1000 1:N:0:CGTGAACTTG
NTTCCAGATATTCGATGCATGTGCCGCTCCTGTCGGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCGTGAT
+
#8BCCGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGG

We have two sequenced DNA fragments (or reads), where each read is represented by blocks of 4 lines,
with sequence and base quality information. Each block of information (4 lines) includes:

1.@ followed by read identifier
2.read sequence
3.+ separator line
4.quality of each base based on ASCII score

Each base has a quality character associated with it, which represents how confidently the machine identified (called) the base. But how does this character translate into a probability of error?
The probability p of error per base is given as a **Phred score**, calculated from a value (Q) derived from
the quality character associated to the base:

$$ p(Q) = 10^(-Q/10) $$

Although there's theoretically no limit, Q usually goes up to around 40 in illumina machines.

To obtain this Q value from the character associated to the quality of the base, we have to know that each character (such as '#') has an ASCII decimal value associated (for example, '#' has a value of 35).
The Q value of a character is the decimal value corresponding to the entry of that character in the ASCII table, subtracted by 33. For example Q('#') = 35 – 33

> ### :pencil2: Hands-on: 
>
> 1. _Interpret the quality of the bases_** : For the fastq file provided initially, analyse some of the bases and obtain their phred score. To do so look at the ASCII table below.
  ![source is www.asciitable.com](http://www.asciitable.com/index/asciifull.gif)
> 2. Step2
>
>    > ### :nut_and_bolt: Comments
>    > A comment
>    {: .comment}
>
>    > ### :bulb: Tip: A tip
>    >
>    > * Step1
>    > * Step2
>    {: .tip}
{: .hands_on}

# Quality Control

High Throughput Sequencing machines read thousands or millions of sequences in paralell. As you can imagine, this usually generates large fastq files, with millions of lines. Manually inspecting the quality of each read is out of the question. Specialized software has been developed to provide quality measures for fastq files generated by HTS machines. FastQC is a popular program to generate quality reports on fastq data. It is usually the first thing you should do once you receive a new dataset.

> ### :pencil2: Hands-on: Data upload
>
> 1. Create a new project
> 2. Step2
>
>    > ### :question: Question
>    >
>    > Question?
>    >
>    > <details>
>    > <summary>Click to view answers</summary>
>    > Answer to question
>    > </details>
>    {: .question}

>    > ### :bulb: Tip: A tip
>    >
>    > * Step1
>    > * Step2
>    {: .tip}
{: .hands_on}

Some blabla
> ### :pencil2: Hands-on: Data upload
>
> 1. Step1
> 2. Step2
>
>    > ### :question: Questions
>    >
>    > 1. Question1?
>    > 2. Question2?
>    >
>    >    <details>
>    >    <summary>Click to view answers</summary>
>    >    <ol type="1">
>    >    <li>Answer for question1</li>
>    >    <li>Answer for question2</li>
>    >    </ol>
>    >    </details>
>    {: .question}
>
> 3. Step3
{: .hands_on}

# Part 2

Short introduction about this subpart.

> ### :nut_and_bolt: Comment
>
> Do you want to learn more about the principles behind mapping? Follow our [training](../../NGS-mapping)
> {: .comment}

# Conclusion
{:.no_toc}

Conclusion about the technical key points. And then relation between the technics and the biological question to end with a global view.
